### PR TITLE
session-service: fix bug causing SAML refresh to fail

### PR DESCRIPTION
**Pull request by srenatus**
...unless there happened to be some old data stored with the session.

### :nut_and_bolt: Description

Before, the "is this a real proper session?" logic had depended on any
key being set in data. For local users and LDAP users, there would be a
`refresh_token`; for SAML users, there wouldn't.

Before #596, SAML users would have their already-used `client_state` and
`relay_state` keys, and their sessions would be non-empty. With that
change, however, we're supporting multiple such keys, and to avoid
accumulating them, they're removed when used. This would lead to SAML
users having empty session data.

Now, we explicitly add an `alive: true` to the session data when a new
session is requested.

### :+1: Definition of Done

See demo script 👇 

### :athletic_shoe: Demo Script / Repro Steps

- `start_all_services`, `rebuild components/session-service`.
- login using SAML
- wait for two minutes
- observe that you haven't been logged out

- if you're keen, check session data in postgres: 
```
[13][default:/src:0]# chef-automate dev psql chef_session_service -- -c "select encode(data, 'escape') from sessions"
                                                             encode
--------------------------------------------------------------------------------------------------------------------------------
 {"data":{"alive":true,"client_state_7dphJNj8-OcRMA==":"/","relay_state_7dphJNj8-OcRMA==":true},"deadline":1561815864283360262}
(1 row)
```

### :chains: Related Resources

- Zendesk ticket 22666

### :white_check_mark: Checklist

- [X] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [X] Code actually executed?
- [X] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/session-service/server/server.go [hunk 1]
```diff
@@ -64,9 +64,10 @@ type Server struct {
 }
 
 const (
+	aliveKey                 = "alive"
 	refreshTokenKey          = "refresh_token"
-	relayStateKey            = "relay_state"  // random, passed to OIDC IdP
-	clientStateKey           = "client_state" // coming from the browser, passed back on success
+	relayStateKeyPrefix      = "relay_state"  // random, passed to OIDC IdP
```
Comment (srenatus) [inline components/session-service/server/server.go]: drive-by: call the prefixes what they are

File: components/session-service/server/server.go [hunk 2]
```diff
@@ -64,9 +64,10 @@ type Server struct {
 }
 
 const (
+	aliveKey                 = "alive"
 	refreshTokenKey          = "refresh_token"
-	relayStateKey            = "relay_state"  // random, passed to OIDC IdP
-	clientStateKey           = "client_state" // coming from the browser, passed back on success
+	relayStateKeyPrefix      = "relay_state_"  // random, passed to OIDC IdP
+	clientStateKeyPrefix     = "client_state_" // coming from the browser, passed back on success
```
Comment (srenatus) [inline components/session-service/server/server.go]: drive-by cleanup

File: components/session-service/server/server.go [hunk 3]
```diff
@@ -454,17 +455,19 @@ func (s *Server) refreshHandler(w http.ResponseWriter, r *http.Request) {
 	// maybeExchangeRefreshTokenForIDToken will return the still-valid idToken.
 
 	// Check if the session exists
-	// Note 2018/01/05 (sr): Couldn't find a better way.
-	if keys, err := sess.Keys(); err != nil { // nolint: vetshadow
+	if alive, err := sess.GetBool(aliveKey); err != nil {
+		s.log.Error("error retrieving session alive key")
 		httpError(w, http.StatusInternalServerError)
 		return
-	} else if len(keys) == 0 {
```
Comment (srenatus) [inline components/session-service/server/server.go]: ⚡️ This is the assumption that was invalidated for SAML users in #596.

File: components/session-service/server/server.go [hunk 4]
```diff
@@ -579,20 +582,27 @@ func (s *Server) newHandler(w http.ResponseWriter, r *http.Request) {
...
-	if err := sess.PutBool(w, relayStateKey+"_"+relayState, true); err != nil {
+	if err := sess.PutBool(w, relayStateKeyPrefix+relayState, true); err != nil {
 		s.log.Errorf("couldn't put relay state into session: %s", err)
 		httpError(w, http.StatusInternalServerError)
 		return
 	}
 	s.log.Debugf("stored relayState %s", relayState)
 	authCodeURL := s.client.AuthCodeURL(relayState)
+
+	if err := sess.PutBool(w, aliveKey, true); err != nil {
```
Comment (srenatus) [inline components/session-service/server/server.go]: This way, we don't rely on having something in session data to conclude that it's a valid session; either it has `alive:true`, or it isn't.

File: components/session-service/server/server_test.go [hunk 5]
```diff
@@ -33,23 +32,23 @@ const bldrURLString = "https://builder.test/"
 
 func TestSafeCmpStrings(t *testing.T) {
 	t.Run("when the strings are the same, it returns true", func(t *testing.T) {
```
Comment (srenatus) [inline components/session-service/server/server_test.go]: drive-by cleanup

File: components/session-service/server/server.go [hunk 6]
```diff
@@ -454,17 +455,19 @@ func (s *Server) refreshHandler(w http.ResponseWriter, r *http.Request) {
...
 
 	// Check if the session exists
-	// Note 2018/01/05 (sr): Couldn't find a better way.
-	if keys, err := sess.Keys(); err != nil { // nolint: vetshadow
+	if alive, err := sess.GetBool(aliveKey); err != nil {
+		s.log.Error("error retrieving session alive key")
 		httpError(w, http.StatusInternalServerError)
 		return
-	} else if len(keys) == 0 {
+	} else if !alive {
```
Comment (blakestier) [inline components/session-service/server/server.go]: maybe I've lost track but does this need to be an `else if`? There's only two states, right?

Comment (blakestier) [reply components/session-service/server/server.go]: not worth blocking on tho

Comment (srenatus) [reply components/session-service/server/server.go]: The if branch is `err! = nil`,  when the thing couldn't be retrieved

Comment (blakestier) [reply components/session-service/server/server.go]: hmm the var `alive` assignment is confusing me. So this reads `if err is not nil, do this, else if it's not alive, do this` and then the rest is if `alive`? No extra `else` required?

Comment (msorens) [reply components/session-service/server/server.go]: Correct @blakestier.
Would it clearer if we were to remove the `else`? (That's valid because the previous block includes a `return`.) Then it is:
```
if err is not nil do this and return
if !alive do that and return
// continue with "alive" case
```
